org.eclipse.xtext.xbase.ui/src/org/eclipse/xtext/xbase/ui/editor/XbaseEditorInputRedirector.java

```diff
@@ -0,0 +1,162 @@
+				// if the given resource is already pointing to a language file
+				if (fileExtensionProvider.isValid(resource.getFullPath().getFileExtension())) {
+					IJavaProject project = JavaCore.create(resource.getProject());
+					if (project.exists()) {
+						// Check if it's sitting in the output dir, look for original one
+						if (project.getOutputLocation().isPrefixOf(resource.getFullPath())) {
+							IPath relative = resource.getFullPath().removeFirstSegments(project.getOutputLocation().segmentCount());
+
+							List<IPackageFragmentRoot> packageFragmentRoots = new ArrayList<>();
+							for (IPackageFragmentRoot root : project.getPackageFragmentRoots()) {
+								if (root.getKind() == IPackageFragmentRoot.K_SOURCE) {
+									packageFragmentRoots.add(root);
+								}
+							}
+							
+							for (IPackageFragmentRoot source : packageFragmentRoots) {
+								IPath fullPath = source.getCorrespondingResource().getProjectRelativePath().append(relative);
+								IFile newFile = resource.getProject().getFile(fullPath);
+								if (newFile.exists()) {
+									return new FileEditorInput(newFile);
+								}
+							}
+						}
+
+						// Check if it's sitting in one of the output folders set on the source folders
+						List<IClasspathEntry> classpathEntries = new ArrayList<>();
+						for (IClasspathEntry classpathEntry : project.getRawClasspath()) {
+							if (classpathEntry.getEntryKind() == IClasspathEntry.CPE_SOURCE) {
+								classpathEntries.add(classpathEntry);
+							}
+						}
+						
+						for (IClasspathEntry sourceFolder : classpathEntries) {
+							if (sourceFolder.getOutputLocation() != null && sourceFolder.getOutputLocation().isPrefixOf(resource.getFullPath())) {
+								IPath relative = resource.getFullPath().removeFirstSegments(sourceFolder.getOutputLocation().segmentCount());
+								Optional<IPackageFragmentRoot> source = FluentIterable.from(project.findPackageFragmentRoots(sourceFolder))
+										.first();
+								if (source.isPresent()) {
+									IPath fullPath = source.get().getCorrespondingResource().getProjectRelativePath().append(relative);
+									IFile newFile = resource.getProject().getFile(fullPath);
+									if (newFile.exists()) {
+										return new FileEditorInput(newFile);
+									}
+								}
+							}
+						}
+
+					}
+				}
+			}
+		} catch (JavaModelException e) {
+			LOG.error("Error finding the original source for the editor input", e);
+		}
+
+		return input;
+	}
+
+	public IEditorInput findOriginalSource(IEditorInput input) {
+		IFile resource = ResourceUtil.getFile(input);
+		if (resource != null) {
+			IEditorInput original = findOriginalSourceForOuputFolderCopy(input);
+			if (original != input) {
+				return original;
+			}
+
+			IEclipseTrace trace = traceInformation.getTraceToSource(resource);
+			if (trace == null) {
+				return input;
+			}
+
+			Optional<? extends ILocationInEclipseResource> sourceInformation = Iterables.tryFind(trace.getAllAssociatedLocations(),
```

Comment:
Simplified the code. The original code made use of a `while-loop` to find the first element for which the “languageInfo.equals(candidate.getLanguage())” is true. I do the same using the new code